syntax = "proto3";

package main;

message Empty {

}

message Runway {
    int32 runway = 1;
    string airportName = 2;
    string preCode = 3;
    int32 height = 4;
}

message ArrivingPlane {
    string srcAirport = 1;
    string code = 2;
    string ip = 3;
    int32 port = 4;
}

message DepartingPlane {
    string code = 1;
    int32 runway = 2;
    int32 height = 3;
    string airportName = 4;
}

message PlaneData {
    string code = 1;
    int32 currFuel = 2;
    int32 maxFuel = 3;
    int32 currWeight = 4;
    int32 maxWeight = 5;
}

message TakeoffStatus {
    int32 errorCode = 1;
}

message Flight {
    string code = 1;
    string time = 2;
    string airport = 3;
    int32 type = 4;
    int32 runway = 5;
}

message planeHeight {
    int32 height = 1;
}

message AirportInfo {
    string ip = 1;
    int32 port = 2;
}

message ScreenInfo {
    string ip = 1;
    int32 port = 2;
}

message AirportName {
    string name = 1;
}

service towerHost {
    rpc requestLanding(ArrivingPlane) returns (Runway);
    rpc requestTakeoff(DepartingPlane) returns (Runway);
    rpc checkTakeoff(PlaneData) returns (TakeoffStatus);
    rpc takeoff (DepartingPlane) returns (AirportInfo);
    rpc screenConnect(ScreenInfo) returns (AirportName);
}

service planeHost {
    rpc notifyLanding(Runway) returns (planeHeight);
    rpc notifyDeparture(Runway) returns (Empty);
}

service screenHost {
    rpc listFlights(stream Flight) returns (Empty) {}
}